### PR TITLE
Yautja Bracer Charging Rework

### PR DESCRIPTION
**Pull request by realforest2001**
# About the pull request
Changes how bracers recharge, requiring drain from APC in the field. Change made as atomisation from #3271 as part of an ongoing project to make Yautja a more balanced role for what it is. 

# Explain why it's good for the game
Balances out power usage and makes things more dynamic. Will tie in with cloak not using power but being more vulnerable to the methods to force decloak.
# Testing Photographs and Procedure
<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog
:cl:
add: Yautja bracers can now be recharged from APCs.
remove: Yautja bracers no longer naturally recharge on the colony.
/:cl:

### CODE REVIEW

File: code/modules/cm_preds/yaut_bracers.dm [hunk 1]
```diff
@@ -31,6 +31,9 @@
 	var/notification_sound = TRUE // Whether the bracer pings when a message comes or not
 	var/charge = 1500
 	var/charge_max = 1500
+	var/charge_rate = 30
+	var/charge_cooldown = 3 MINUTES
+	var/last_charge
```
Comment (BeagleGaming1) [inline code/modules/cm_preds/yaut_bracers.dm:36]: code doc for the vars?

File: code/modules/cm_preds/yaut_bracers.dm [hunk 2]
```diff
@@ -75,11 +78,27 @@
 	if(!ishuman(loc))
 		STOP_PROCESSING(SSobj, src)
 		return
-	var/mob/living/carbon/human/H = loc
+	var/mob/living/carbon/human/hooman = loc
```
Comment (BeagleGaming1) [inline code/modules/cm_preds/yaut_bracers.dm:81]: why "hooman"?

File: code/modules/power/apc.dm [hunk 3]
```diff
@@ -798,41 +798,87 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
...
 					return
 
-				var/obj/item/storage/backpack/marine/smartpack/S = H.back
-				if(S.battery_charge < SMARTPACK_MAX_POWER_STORED)
-					var/charge_to_use = min(cell.charge, SMARTPACK_MAX_POWER_STORED - S.battery_charge)
-					if(!(cell.use(charge_to_use)))
+				playsound(src.loc, 'sound/effects/sparks2.ogg', 25, 1)
+
+				if(stat & BROKEN)
+					var/datum/effect_system/spark_spread/s = new /datum/effect_system/spark_spread
```
Comment (BeagleGaming1) [inline code/modules/power/apc.dm:817]: single letter var

File: code/modules/power/apc.dm [hunk 4]
```diff
@@ -798,41 +798,87 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
...
+					return FALSE
+				if(bracer.last_charge > world.time)
+					to_chat(user, SPAN_WARNING("It is too soon for [bracer.name] to siphon power again."))
+					return FALSE
+				to_chat(user, SPAN_NOTICE("You rest your bracer against the APC interface and begin to siphon off some of the stored energy."))
+				if(!do_after(hooman, 20, INTERRUPT_ALL, BUSY_ICON_HOSTILE))
+					return FALSE
+
+				if(stat & BROKEN)
+					var/datum/effect_system/spark_spread/s = new /datum/effect_system/spark_spread
```
Comment (BeagleGaming1) [inline code/modules/power/apc.dm:852]: single letter var

**Review by BeagleGaming1 — CHANGES_REQUESTED** [review]

File: code/modules/power/apc.dm [hunk 5]
```diff
@@ -798,41 +798,87 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
...
+					return FALSE
+				if(bracer.last_charge > world.time)
+					to_chat(user, SPAN_WARNING("It is too soon for [bracer.name] to siphon power again."))
+					return FALSE
+				to_chat(user, SPAN_NOTICE("You rest your bracer against the APC interface and begin to siphon off some of the stored energy."))
+				if(!do_after(grabber, 20, INTERRUPT_ALL, BUSY_ICON_HOSTILE))
+					return FALSE
+
+				if(stat & BROKEN)
+					var/datum/effect_system/spark_spread/spark = new /datum/effect_system/spark_spread
```
Comment (harryob) [inline code/modules/power/apc.dm:852]:
```suggestion
					var/datum/effect_system/spark_spread/spark = new()
```

File: code/modules/power/apc.dm [hunk 6]
```diff
@@ -798,41 +798,87 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
...
+					spark.set_up(3, 1, src)
+					spark.start()
+					to_chat(grabber, SPAN_DANGER("The APC's power currents surge eratically, super-heating your bracer!"))
+					playsound(src.loc, 'sound/effects/sparks2.ogg', 25, 1)
+					grabber.apply_damage(10,0, BURN)
+					return FALSE
+				else if(cell && cell.charge > 0)
```
Comment (harryob) [inline code/modules/power/apc.dm:859]: redundant
```suggestion

				if(cell && cell.charge > 0)
```

File: code/modules/power/apc.dm [hunk 7]
```diff
@@ -798,41 +798,87 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
...
+						charging = APC_CHARGING
+						set_broken() // Breaks the APC
+
+						return TRUE
+					else
+						to_chat(user, SPAN_WARNING("[bracer.name] is already fully charged."))
+						return FALSE
+				else
+					to_chat(user, SPAN_WARNING("There is no charge to draw from that APC."))
+					return FALSE
```
Comment (harryob) [inline code/modules/power/apc.dm:879]:
```suggestion
				if(!cell || !cell.charge <= 0)
					to_chat(user, SPAN_WARNING("There is no charge to draw from that APC."))
					return FALSE

				if(bracer.charge_max <= bracer.charge)
					to_chat(user, SPAN_WARNING("[bracer.name] is already fully charged."))
					return FALSE

				var/charge_to_use = min(cell.charge, bracer.charge_max - bracer.charge)
				if(!(cell.use(charge_to_use)))
					return FALSE
				playsound(src.loc, 'sound/effects/sparks2.ogg', 25, 1)
				bracer.charge += charge_to_use
				bracer.last_charge = (world.time + bracer.charge_cooldown)
				to_chat(grabber, SPAN_YAUTJABOLD("[icon2html(bracer)] \The <b>[bracer]</b> beep: Power siphon complete. Charge at [bracer.charge]/[bracer.charge_max]."))
				if(bracer.notification_sound)
					playsound(bracer.loc, 'sound/items/pred_bracer.ogg', 75, 1)
				charging = APC_CHARGING
				set_broken() // Breaks the APC

				return TRUE
```

File: code/modules/power/apc.dm [hunk 8]
```diff
@@ -1011,9 +1057,9 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
 			smoke.set_up(1, 0, loc)
 			smoke.attach(src)
 			smoke.start()
-			var/datum/effect_system/spark_spread/s = new /datum/effect_system/spark_spread
-			s.set_up(1, 1, src)
-			s.start()
+			var/datum/effect_system/spark_spread/spark = new /datum/effect_system/spark_spread
```
Comment (harryob) [inline code/modules/power/apc.dm:1060]:
```suggestion
			var/datum/effect_system/spark_spread/spark = new()
```

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/cm_preds/yaut_bracers.dm [hunk 9]
```diff
@@ -31,6 +31,12 @@
 	var/notification_sound = TRUE // Whether the bracer pings when a message comes or not
 	var/charge = 1500
 	var/charge_max = 1500
+	/// The amount charged per process
+	var/charge_rate = 30
+	/// Cooldown on draining power from APC
+	var/charge_cooldown = 3 MINUTES
+	/// Worldtime of last recharge
+	var/last_charge
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:39]: use the cooldown defines for managed cooldowns, either COOLDOWN_ or TIMER_COOLDOWN_

File: code/modules/cm_preds/yaut_bracers.dm [hunk 10]
```diff
@@ -75,11 +81,27 @@
 	if(!ishuman(loc))
 		STOP_PROCESSING(SSobj, src)
 		return
-	var/mob/living/carbon/human/H = loc
+	var/mob/living/carbon/human/hooman = loc
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:84]:
```suggestion
	var/mob/living/carbon/human/human_holder = loc
```

File: code/modules/cm_preds/yaut_bracers.dm [hunk 11]
```diff
@@ -75,11 +81,27 @@
...
 
-	charge = min(charge + 30, charge_max)
-	var/perc_charge = (charge / charge_max * 100)
-	H.update_power_display(perc_charge)
+	if(charge < charge_max)
+		var/charge_increase = charge_rate
+		if(is_ground_level(hooman.z))
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:88]:
```suggestion
		if(is_ground_level(human_holder.z))
```

File: code/modules/cm_preds/yaut_bracers.dm [hunk 12]
```diff
@@ -75,11 +81,27 @@
...
+			charge_increase = charge_rate / 6
+		else if(is_mainship_level(hooman.z))
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:90]:
```suggestion
		else if(is_mainship_level(human_holder.z))
```

File: code/modules/cm_preds/yaut_bracers.dm [hunk 13]
```diff
@@ -75,11 +81,27 @@
...
+			charge_increase = charge_rate / 3
+
+		charge = min(charge + charge_increase, charge_max)
+		var/perc_charge = (charge / charge_max * 100)
+		hooman.update_power_display(perc_charge)
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:95]:
```suggestion
		human_holder.update_power_display(perc_charge)
```

File: code/modules/cm_preds/yaut_bracers.dm [hunk 14]
```diff
@@ -75,11 +81,27 @@
...
+
+	//Non-Yautja have a chance to get stunned with each power drain
+	if(cloaked)
+		if(!HAS_TRAIT(hooman, TRAIT_YAUTJA_TECH) && !hooman.hunter_data.thralled)
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:99]:
```suggestion
		if(!HAS_TRAIT(human_holder, TRAIT_YAUTJA_TECH) && !human_holder.hunter_data.thralled)
```

File: code/modules/cm_preds/yaut_bracers.dm [hunk 15]
```diff
@@ -75,11 +81,27 @@
...
+			if(prob(15))
+				decloak(hooman)
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:101]:
```suggestion
				decloak(human_holder)
```

File: code/modules/cm_preds/yaut_bracers.dm [hunk 16]
```diff
@@ -75,11 +81,27 @@
...
+				shock_user(hooman)
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:102]:
```suggestion
				shock_user(human_holder)
```

File: code/modules/cm_preds/yaut_bracers.dm [hunk 17]
```diff
@@ -75,11 +81,27 @@
...
+		if(hooman.stat == DEAD)
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:103]:
```suggestion
		if(human_holder.stat == DEAD)
```

File: code/modules/cm_preds/yaut_bracers.dm [hunk 18]
```diff
@@ -75,11 +81,27 @@
...
+			decloak(hooman, TRUE)
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:104]:
```suggestion
			decloak(human_holder, TRUE)
```

File: code/modules/power/apc.dm [hunk 19]
```diff
@@ -798,41 +798,87 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
...
+					to_chat(user, SPAN_WARNING("[bracer.name] is already fully charged."))
+					return FALSE
+
+				var/charge_to_use = min(cell.charge, bracer.charge_max - bracer.charge)
+				if(!(cell.use(charge_to_use)))
+					return FALSE
+				playsound(src.loc, 'sound/effects/sparks2.ogg', 25, 1)
+				bracer.charge += charge_to_use
+				bracer.last_charge = (world.time + bracer.charge_cooldown)
+				to_chat(grabber, SPAN_YAUTJABOLD("[icon2html(bracer)] \The <b>[bracer]</b> beep: Power siphon complete. Charge at [bracer.charge]/[bracer.charge_max]."))
```
Comment (harryob) [inline code/modules/power/apc.dm:873]:
```suggestion
				to_chat(grabber, SPAN_YAUTJABOLD("[icon2html(bracer)] <b>[bracer]</b> beeps: Power siphon complete. Charge at [bracer.charge]/[bracer.charge_max]."))
```

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/power/apc.dm (hunk 19 above)
Comment (realforest2001) [reply code/modules/power/apc.dm:873]: They're pluralised bracerS 😔

File: code/modules/power/apc.dm [hunk 20]
```diff
@@ -798,41 +798,87 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
...
+
+				if(stat & BROKEN)
+					var/datum/effect_system/spark_spread/spark = new()
+					spark.set_up(3, 1, src)
+					spark.start()
+					to_chat(grabber, SPAN_DANGER("The APC's power currents surge eratically, super-heating your bracer!"))
+					playsound(src.loc, 'sound/effects/sparks2.ogg', 25, 1)
+					grabber.apply_damage(10,0, BURN)
+					return FALSE
+				if(!cell || !cell.charge <= 0)
```
Comment (morrowwolf) [inline code/modules/power/apc.dm:859]: if(!cell || cell.charge <= 0)

**Review by morrowwolf — CHANGES_REQUESTED** [review]

File: code/modules/power/apc.dm (hunk 20 above)
Comment (realforest2001) [inline code/modules/power/apc.dm:859]:
```suggestion
				if(!cell || cell.charge <= 0)
```

File: code/modules/cm_preds/yaut_bracers.dm [hunk 21]
```diff
@@ -75,11 +81,27 @@
...
+		if(is_ground_level(human_holder.z))
+			charge_increase = charge_rate / 6
+		else if(is_mainship_level(human_holder.z))
+			charge_increase = charge_rate / 3
+
+		charge = min(charge + charge_increase, charge_max)
+		var/perc_charge = (charge / charge_max * 100)
+		human_holder.update_power_display(perc_charge)
+
+	//Non-Yautja have a chance to get stunned with each power drain
```
Comment (Diegoflores31) [inline code/modules/cm_preds/yaut_bracers.dm:97]: can research trait users get a little more leeway ?

**Comment by Mister-moon1** [comment]
During the original testing for the pred quality of life TM it was expressed by many that achieving power planetside was, although difficult, possible due to a couple APCs missed by the xenos.
The noticeable oversight was of course the fact of having near enough infinite power shipside due to being able to ping to and from the same APC, the "solution" given was to have the APC destroyed upon siphon along with a 5 minute cooldown

This has led to achieving power groundside almost impossible via siphoning power due to the fact of that one forgotten APC just being destroyed when the yautja gets power, leaving no source of power other than the 5 power per second or so from natural regeneration, when needing 1000 for a simple crystal this is a bit absurd. It also leaves the only other possible option to go shipside and deal with the better but still subpar power regeneration with the best solution to go afk and come back when it is done which i do not find very engaging.

I suggest that the groundside regeneration be doubled to 10 per tick and APCs be reverted to their state of not being broken upon siphon, i believe that the 5 minute cooldown should be enough to not warrant issues coming from this

**Comment by Diegoflores31** [comment]
Idea for "possible" solution . Add a self recharge device thingy that can be used like like AVP classic one . 10 minute cooldown or more . With a windup of dunno 10 seconds

**Comment by morrowwolf** [comment]
Ping me once cloaking PR is merged if I forget.

File: code/modules/power/apc.dm [hunk 22]
```diff
@@ -798,41 +798,87 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
...
+				if(bracer.charge_max <= bracer.charge)
+					to_chat(user, SPAN_WARNING("[bracer.name] is already fully charged."))
+					return FALSE
+
+				var/charge_to_use = min(cell.charge, bracer.charge_max - bracer.charge)
+				if(!(cell.use(charge_to_use)))
+					return FALSE
+				playsound(src.loc, 'sound/effects/sparks2.ogg', 25, 1)
+				bracer.charge += charge_to_use
+				bracer.last_charge = (world.time + bracer.charge_cooldown)
```
Comment (harryob) [inline code/modules/power/apc.dm:872]: cooldown defines

File: code/modules/cm_preds/yaut_bracers.dm (hunk 21 above)
Comment (harryob) [reply code/modules/cm_preds/yaut_bracers.dm:97]: no

File: code/modules/cm_preds/yaut_bracers.dm [hunk 23]
```diff
@@ -31,6 +31,12 @@
 	var/notification_sound = TRUE // Whether the bracer pings when a message comes or not
 	var/charge = 1500
 	var/charge_max = 1500
+	/// The amount charged per process
+	var/charge_rate = 30
+	/// Cooldown on draining power from APC
+	var/charge_cooldown = COOLDOWN_BRACER_CHARGE
+	/// Worldtime of last recharge
+	var/last_charge
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:39]: cooldown defines

File: code/modules/cm_preds/yaut_bracers.dm [hunk 24]
```diff
@@ -75,11 +81,27 @@
...
+			charge_increase = charge_rate / 6
+		else if(is_mainship_level(human_holder.z))
+			charge_increase = charge_rate / 3
+
+		charge = min(charge + charge_increase, charge_max)
+		var/perc_charge = (charge / charge_max * 100)
+		human_holder.update_power_display(perc_charge)
+
+	//Non-Yautja have a chance to get stunned with each power drain
+	if(cloaked)
```
Comment (harryob) [inline code/modules/cm_preds/yaut_bracers.dm:98]: early return

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/power/apc.dm [hunk 25]
```diff
@@ -798,41 +798,87 @@ GLOBAL_LIST_INIT(apc_wire_descriptions, list(
...
 					return
 
-				var/obj/item/storage/backpack/marine/smartpack/S = H.back
-				if(S.battery_charge < SMARTPACK_MAX_POWER_STORED)
-					var/charge_to_use = min(cell.charge, SMARTPACK_MAX_POWER_STORED - S.battery_charge)
-					if(!(cell.use(charge_to_use)))
+				playsound(src.loc, 'sound/effects/sparks2.ogg', 25, 1)
+
+				if(stat & BROKEN)
+					var/datum/effect_system/spark_spread/spark = new /datum/effect_system/spark_spread
```
Comment (harryob) [inline code/modules/power/apc.dm:817]:
```suggestion
					var/datum/effect_system/spark_spread/spark = new()
```

**Review by harryob — CHANGES_REQUESTED** [review]
one last nit

**Review by harryob — APPROVED** [review]

**Pull request merged**